src/backend/distributed/commands/create_distributed_table.c

```diff
@@ -146,199 +162,417 @@ master_create_distributed_table(PG_FUNCTION_ARGS)
+{
 	Oid relationId = PG_GETARG_OID(0);
-	text *distributionColumnText = PG_GETARG_TEXT_P(1);
-	Oid distributionMethodOid = PG_GETARG_OID(2);
 
-	char *distributionColumnName = text_to_cstring(distributionColumnText);
-	char distributionMethod = LookupDistributionMethod(distributionMethodOid);
-	text *colocateWithTableNameText = NULL;
+	Relation relation = NULL;
 	char *colocateWithTableName = NULL;
+	List *workerNodeList = NIL;
+	int workerCount = 0;
+	Var *distributionColumn = NULL;
+
+	bool viaDeprecatedAPI = false;
 
 	EnsureCoordinator();
 	CheckCitusVersion(ERROR);
 
-	/* guard against a binary update without a function update */
-	if (PG_NARGS() >= 4)
+	/*
+	 * Lock target relation with an exclusive lock - there's no way to make
+	 * sense of this table until we've committed, and we don't want multiple
+	 * backends manipulating this relation.
+	 */
+	relation = relation_open(relationId, ExclusiveLock);
+
+	/*
+	 * We should do this check here since the codes in the following lines rely
+	 * on this relation to have a supported relation kind. More extensive checks
+	 * will be performed in CreateDistributedTable.
+	 */
+	EnsureRelationKindSupported(relationId);
+
+	workerNodeList = ActivePrimaryNodeList();
+	workerCount = list_length(workerNodeList);
+
+	/* if there are no workers, error out */
+	if (workerCount == 0)
+	{
+		char *relationName = get_rel_name(relationId);
+
+		ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+						errmsg("cannot create reference table \"%s\"", relationName),
+						errdetail("There are no active worker nodes.")));
+	}
+
+
+	CreateDistributedTable(relationId, distributionColumn, DISTRIBUTE_BY_NONE,
+						   colocateWithTableName, viaDeprecatedAPI);
+
+	relation_close(relation, NoLock);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * CreateDistributedTable creates distributed table in the given configuration.
+ * This functions contains all necessary logic to create distributed tables. It
+ * perform necessary checks to ensure distributing the table is safe. If it is
+ * safe to distribute the table, this function creates distributed table metadata,
+ * creates shards and copies local data to shards.
+ *
+ * viaDeprecatedAPI boolean flag is not optimal way to implement this function,
+ * but it helps reducing code duplication a lot. We hope to remove that flag one
+ * day. once we deprecate master_create_distribute_table completely.
+ */
+void
+CreateDistributedTable(Oid relationId, Var *distributionColumn, char distributionMethod,
+					   char *colocateWithTableName, bool viaDeprecatedAPI)
+{
+	char replicationModel = REPLICATION_MODEL_INVALID;
+	uint32 colocationId = INVALID_COLOCATION_ID;
+	Oid colocatedTableId = InvalidOid;
+	bool localTableEmpty = false;
+
+	Relation colocatedRelation = NULL;
+
+	replicationModel = AppropriateReplicationModel(distributionMethod, viaDeprecatedAPI);
+
+	colocationId = ColocationIdForNewTable(relationId, distributionColumn,
```

Comment:
`ColocationIdForNewTable ` assumes that the caller acquires the lock. But, the upper level caller acquired the lock in this case. Shall we add a single line comment saying this ?